common/setups/rasr/nn_system.py

```diff
@@ -0,0 +1,554 @@
+        self.dev_input_data = dev_data
+        self.test_input_data = test_data
+
+        self.train_corpora.extend(list(train_data.keys()))
+        self.cv_corpora.extend(list(cv_data.keys()))
+        self.devtrain_corpora.extend(list(devtrain_data.keys()))
+        self.dev_corpora.extend(list(dev_data.keys()))
+        self.test_corpora.extend(list(test_data.keys()))
+
+        self.train_cv_pairing = train_cv_pairing
+
+        for trn_c, cv_c in itertools.product(self.train_corpora, self.cv_corpora):
+            if (trn_c, cv_c) in self.train_cv_pairing or self.train_cv_pairing is None:
+                continue
+
+            self.jobs[f"{trn_c}_{cv_c}"] = {}
+            self.nn_models[f"{trn_c}_{cv_c}"] = {}
+            self.nn_configs[f"{trn_c}_{cv_c}"] = {}
+
+    def prepare_data(self, raw_sampling_rate: int, feature_sampling_rate: int):
+        for name in self.train_corpora + self.devtrain_corpora + self.cv_corpora:
+            self.jobs[name]["ogg_zip"] = j = returnn.BlissToOggZipJob(
+                bliss_corpus=self.crp[name].corpus_config.corpus_file,
+                segments=self.crp[name].segment_path,
+                rasr_cache=self.feature_flows[name]["init"],
+                raw_sample_rate=raw_sampling_rate,
+                feat_sample_rate=feature_sampling_rate,
+            )
+            self.oggzips[name] = j.out_ogg_zip
+            j.add_alias(f"oggzip/{name}")
+
+            # TODO self.jobs[name]["hdf_full"] = j = returnn.ReturnnDumpHDFJob()
+
+    def generate_lattices(self):
+        pass
+
+    # -------------------- Training --------------------
+
+    def returnn_training(
+        self,
+        name,
+        returnn_config,
+        nn_train_args,
+        train_corpus_key,
+        cv_corpus_key,
+        devtrain_corpus_key=None,
+    ):
+        assert isinstance(returnn_config, returnn.ReturnnConfig)
+
+        returnn_config.config["train"] = self.train_input_data[
+            train_corpus_key
+        ].get_data_dict()
+        returnn_config.config["dev"] = self.cv_input_data[cv_corpus_key].get_data_dict()
+        if devtrain_corpus_key is not None:
+            returnn_config.config["eval_datasets"] = {
+                "devtrain": self.devtrain_input_data[
+                    devtrain_corpus_key
+                ].get_data_dict()
+            }
+
+        train_job = returnn.ReturnnTrainingJob(
+            returnn_config=returnn_config,
+            **nn_train_args,
+        )
+
+        train_job.add_alias(f"train_nn/{train_corpus_key}_{cv_corpus_key}/{name}_train")
+        self.jobs[f"{train_corpus_key}_{cv_corpus_key}"][name] = train_job
+        self.nn_models[f"{train_corpus_key}_{cv_corpus_key}"][
+            name
+        ] = train_job.out_models
+        self.nn_configs[f"{train_corpus_key}_{cv_corpus_key}"][
+            name
+        ] = train_job.out_returnn_config_file
+
+        return train_job
+
+    def returnn_rasr_training(
```

Comment:
This function shares almost the same signature as `returnn_training` and does very similar things. 
Would it be possible / desirable to only have one function, which then internally decides on `ReturnnTrainingJob` vs `ReturnnRasrTrainingJob` and does the handling based on `*_corpus_key`.

Reply:
maybe not sure :)